### PR TITLE
[astro] Various refactoring/fixes

### PR DESCRIPTION
**Pull request by Nadahar**
The default `Locale` is determined by the JVM (often deducted from the OS settings, but fundamentally unpredictable), and ideally should never be used. For test in particular, it can easily lead to "unstable" tests.

It seems like `GregorianCalendar` (and probably the other `Calendar` implementation) are very poorly designed, it only has a single constructor where you can set the `Locale`. For all the others, the default `Locale` is used. As such, all use of the class should be restricted to using this one constructor.

This in itself is a much bigger task, I've only addressed on test class in the Astro binding here, based on [this post](https://community.openhab.org/t/working-on-add-on-trying-to-update-from-4-3-3-to-5-1-x-and-cant-build/166176/4). Since I can't reproduce the test failure, I can't know if this will fix it, but I'd say that restricting the test to using "the one good constructor" can't be a bad thing in any case.

### CODE REVIEW

**Comment by Nadahar** [comment]
Are there still dependency problems out there? It doesn't seem like the test failures have anything to do with my changes.

**Comment by jlaur** [comment]
> Are there still dependency problems out there? It doesn't seem like the test failures have anything to do with my changes.

I think it will be fixed by #19311.

**Comment by Nadahar** [comment]
@lsiepel Any idea what's going on here? The binding doesn't fail when building locally, yet it does here. And, the errors don't make much sense to me, it looks like some kind of Bnd/dependency issue, but it's a bit strange that this happens exactly in the binding I've modified.

edit: I'll wait for #19311 and see if that helps.

**Comment by Nadahar** [comment]
After looking a bit deeper, I don't think this will fix the test (it won't harm it either, but there are just too many problems here).

Basically, any methods that use the default timezone or locale are "bugs" when you want to calculate things like this. `Calendar.getInstance()` is just one example of such a method. When searching the binding, I find 14 places that call is used alone.

When I tried to correct just one of those places (`DateTimeUtils.toCalendar()`), I got 28 errors because the timezone and locale parameters are missing (obviously). I'm deeming this too much work, it seems like most of the code has been written without taking this into account.

Some of the tests that fails are "next season" calculations that include roll over to a new year, using "edge case" values. I see from the stack trace posted that the seasons are calculated using timezone `America/Chicago`. When I test it, `Europe/Oslo` is used instead.  The time zone in the test is set to `Europe/Amsterdam`, so that's what it's supposed to use, but that's only used for the initial value, further calculations use the JVM default timezone. It's not surprising that you get errors when you make calculations where some `Calendar` instances are `Europe/Amsterdam` while others are `America/Chicago`, and you try to fix edge cases. It's not surprising that it works for me with, since `Europe/Amsterdam` and `Europe/Oslo` is in fact the same timezone.

What I don't quite understand is why this usually works with GitHub's builds, because I assume they have different timezones set.

Making the tests fail is easy, if I for example do this:
```java
    @Test
    public void testGetSeasonSydney() {
        TimeZone.setDefault(TimeZone.getTimeZone("America/Chicago"));
        final Season season = seasonCalc.getSeason(DEC_10_2020, SYDNEY_LATITUDE, true);

        assertNextSeason(season.getAutumn(), 2020, JAN_20_2020, season);
        assertNextSeason(season.getWinter(), 2020, MAY_20_2020, season);
        assertNextSeason(season.getSummer(), 2020, SEPT_20_2020, season);
        assertNextSeason(seasonCalc.getSeason(DEC_10_2021, SYDNEY_LATITUDE, true).getAutumn(), 2021, DEC_10_2020,
                season);
    }
```
..the test will fail.

The binding will probably work correctly most of the time, because the default timezone is usually set to match the location where the OH server is running. Ideally, the binding should use the locale and timezone that is configured under "Regional Settings" in OH's settings. For that to work, these parameters would need to be passed around and used everywhere where `Calendar` instances are created.

This wouldn't be a very difficult task, but quite a few pieces of code must be modified, so it is some work.

Since I guess that usually, the OH server's OS timezone corresponds with the OH configured timezone (and locale), most users won't notice these bugs. But tests will fail "randomly" based on the timezone of the computer doing the build. As such, I see the tests as "the most pressing problem". Maybe the best course of action would simply be to disable these tests until somebody has gone through and "fixed" the binding?

**Comment by Nadahar** [comment]
Maybe this should become an issue?

**Comment by lsiepel** [comment]
I also stumbled across this a few years ago when I did some work on this binding. At the time I didn’t fully understand what was going on, not sure if I do now. Anyway I can imagine creating an issue is the best if you don’t have the time now to fix.

**Comment by Nadahar** [comment]
It's not so much that I don't have the time, it's more that I don't have enough overview of how the binding works to make these changes. You have to make changes pretty much everywhere, and to decide *how* things should be changed (where to get the timezone and locale info from etc.), you should know how things are organized.

But, I'm giving it a go, without the necessary overview. I don't feel confident that I make the correct decisions though, so it will need some serious testing and review.

**Comment by lsiepel** [comment]
I do have some u see standing of this binding. But not an expert. Feel free to ask, together we can make it work 
Thankfully there are also tests

**Comment by Nadahar** [comment]
> I do have some u see standing of this binding. But not an expert. Feel free to ask, together we can make it work
> Thankfully there are also tests

I don't have any questions as such at this time, I have tried to implement timezone and locale everywhere (but I still might have missed some places). However, right now, some of the tests fail because the distance to the moon calculation is off by 5 km. It's quite hard to figure out why these calculations output a slight error unless you really study all the included math. I don't intend to fully familiarize myself with the math involved, and as such I'm "not comfortable" with changing things that I don't have a proper overview of.

I'm hoping that e.g. the 5 km error can be caused by a timezone error, it might make sense that an hour off could result in that, I really don't know. But, these kinds of problems are hard to pin down without knowing the math.

**Comment by lsiepel** [comment]
For my understanding, is this PR complete or do you plan to fix more timezone issues?

**Comment by Nadahar** [comment]
> For my understanding, is this PR complete or do you plan to fix more timezone issues?

The PR does what it initially was made to do, but it doesn't fix the test failure or the other issues I've since discovered. I might as well add those fixes to this PR.

I have done much more than I have pushed, but I still have some issues I haven't figured out. I spent most of yesterday trying to figure out SonaType's "new repo system", so there was no progress on this binding.

I'm converting it to a draft for now.

**Comment by lsiepel** [comment]
Do you have the draft somewhere with the 5km difference, I can try to look k to it this weekend.

**Comment by Nadahar** [comment]
> Do you have the draft somewhere with the 5km difference, I can try to look k to it this weekend.

I haven't pushed it anywhere yet, I could push it here, but I now have your Chromecast branch checked out instead 😉 I haven't looked at this anymore because of the Chromecast stuff. The feeling is that I should wrap this up and let the Chromecast wait, because this is most likely a much smaller undertaking.

I think maybe I was onto something when I last worked on this, I found that the Julian calendar value used during the calculations, when converted back to Gregorian using an online converted, was off by 90 minutes or so. It's a strange number that doesn't correspond with the time zone difference alone, but maybe 60 of those 90 minutes are caused by a time zone - and that it's enough to throw the calculation out of "acceptable range". I don't know, but whatever the reason for the problem, I should try to get to the source of those 90 minutes.

**Comment by Nadahar** [comment]
@lsiepel I now pushed exactly what I have, with testing comments and all. It wasn't meant to be pushed anywhere this way, but if you want to scratch your head a bit, please have at it.

I'm basically trying to find the needle in the moon distance calculation 😉

**Comment by lsiepel** [comment]
> @lsiepel I now pushed exactly what I have, with testing comments and all. It wasn't meant to be pushed anywhere this way, but if you want to scratch your head a bit, please have at it.
> 
> I'm basically trying to find the needle in the moon distance calculation 😉

Will look into it the next few days, currently working on some zwave changes. I left the chromecast for a moment.

**Comment by Nadahar** [comment]
> I left the chromecast for a moment.

I might be better if I switched back to the Astro binding and finished that before working further on the Chromecast stuff as well, as I'm pretty sure that figuring out the problem with the Astro binding is astronomically less work than figuring out the Chromecast stuff.

**Comment by Nadahar** [comment]
I've decided to actually make a change that isn't merely a bugfix here, I've changed the `DateTimeUtils.toCalendar(julianDate)` so that it returns a milliseconds accurate conversion instead of rounded to the closest minute. I've verified the conversion with online tools, and it is precise - none of the tests fails with the change, and I simply have no idea why it was truncated in the first place.

There are other tools in `DateTimeUtils` that lets you truncate to midnight, minute and seconds, so I consider it better if those are used when truncation is needed, to doing a "hidden truncation" in a conversion method that just introduces inaccuracy.

See here for further info: https://github.com/openhab/openhab-addons/pull/11327#discussion_r2347561949

**Comment by Nadahar** [comment]
I've solved the moon distance and position test failures I experienced. It turns out that it was merely because the test targets were too imprecise, as soon as I truncated the times to the minute, the tests passed.

**Comment by Nadahar** [comment]
I found and fixed another fragile test. It would fail if the seconds of the current clock were less than 4, all second values >= 4 would make the test pass.

To elaborate, several places in the tests, a "now" value is retrieved, and then date, hour, minute etc. are set. Seconds and milliseconds are often left as they were from the "now" value, i.e. completely random. Often, this doesn't make a difference for the test, but sometimes, in edge cases, this actually matters. My "fix" has mostly been to set seconds and milliseconds to zero, removing the "random factor".

**Comment by Nadahar** [comment]
This PR has evolved to be quite a lot more than I first intended. Most of it is fixes, but there are a few changes to behavior that might need to be discussed. I've made relatively detailed commit messages, so I believe they should provide a relatively decent description of what was changed.

I've not been able to make the tests fail using different timezones or being run at particular times (one used to test if the second was between 0 and 3 at the time it was run). There might be things I've missed, but I've tried to remove "sources of randomness" where I've found them.

If there are any questions, comments or a desire for discussion about certain points, please go ahead and I'll answer as best I can.

**Comment by Nadahar** [comment]
@lsiepel I've taken this out of draft again, I think it's ready now. Also, I forgot to mention, but apart from the unstable tests, it doesn't really fix anything "critical", so I don't think there is any need for backporting.

**Comment by Nadahar** [comment]
I'll add that I realize that this can be a lot to review, because things have been changed a little bit here and a little bit there all over the place. I'd recommend reviewing it on a per-commit basis, I think things will make a lot more sense that way.

File: bundles/org.openhab.binding.astro/src/test/java/org/openhab/binding/astro/internal/util/DateTimeUtilsTest.java [hunk 1]
```diff
@@ -85,12 +87,14 @@ private static void assertNextSeason(Calendar expectedSeason, int expectedYear,
...
+        assertNotNull(nextSeason);
         assertEquals(expectedYear, nextSeason.get(Calendar.YEAR), "Should return the year matching the next season.");
     }
 
     private static Calendar newCalendar(int year, int month, int dayOfMonth, int hourOfDay, int minute, TimeZone zone) {
-        Calendar result = new GregorianCalendar(year, month, dayOfMonth, hourOfDay, minute);
-        result.setTimeZone(zone);
+        Calendar result = new GregorianCalendar(zone, Locale.ROOT);
+        result.set(Calendar.MILLISECOND, 0);
+        result.set(year, month, dayOfMonth, hourOfDay, minute, 0);
```
[the remark this answers is not in the document]
Comment (Nadahar) [reply bundles/org.openhab.binding.astro/src/test/java/org/openhab/binding/astro/internal/util/DateTimeUtilsTest.java:97]: This is a test and is intended.

[the remark this answers is not in the document]
Comment (Nadahar) [reply bundles/org.openhab.binding.astro/src/test/java/org/openhab/binding/astro/internal/util/DateTimeUtilsTest.java:97]: This is intended.

File: bundles/org.openhab.binding.astro/src/test/java/org/openhab/binding/astro/internal/calc/SunCalcTest.java [hunk 2]
```diff
@@ -326,25 +436,26 @@ public void testIssue7642DaylightStart() {
      * @return
      */
     private static Calendar newCalendar(int year, int month, int dayOfMonth, int hourOfDay, int minute, TimeZone zone) {
-        Calendar result = new GregorianCalendar(year, month, dayOfMonth, hourOfDay, minute);
-        result.setTimeZone(zone);
+        Calendar result = new GregorianCalendar(zone, Locale.ROOT);
+        result.set(year, month, dayOfMonth, hourOfDay, minute);
```
[the remark this answers is not in the document]
Comment (Nadahar) [reply bundles/org.openhab.binding.astro/src/test/java/org/openhab/binding/astro/internal/calc/SunCalcTest.java:440]: Intended - It's truncated to minute in the next line, which also resets milliseconds.

File: bundles/org.openhab.binding.astro/src/test/java/org/openhab/binding/astro/internal/calc/MoonCalcTest.java [hunk 3]
```diff
@@ -182,9 +200,9 @@ public void testGetMoonInfoForMoonPhaseAccuracy() {
      * @return
      */
     private static Calendar newCalendar(int year, int month, int dayOfMonth, int hourOfDay, int minute, TimeZone zone) {
-        Calendar result = new GregorianCalendar(year, month, dayOfMonth, hourOfDay, minute);
-        result.setTimeZone(zone);
+        Calendar result = new GregorianCalendar(zone, Locale.ROOT);
+        result.set(year, month, dayOfMonth, hourOfDay, minute);
```
[the remark this answers is not in the document]
Comment (Nadahar) [reply bundles/org.openhab.binding.astro/src/test/java/org/openhab/binding/astro/internal/calc/MoonCalcTest.java:204]: Intended - It's truncated to minute in the next line, which also resets milliseconds.

**Comment by Nadahar** [comment]
@lsiepel All comments from Copilot have been dismissed.

**Comment by Nadahar** [comment]
I've never seen Copilot not generate a description before, but it's not surprising really, since despite all the changes to the code, this PR isn't intended to change the behavior of the binding - except for removing bugs and the increase in conversion accuracy when converting from Julian Calendar.

**Comment by Nadahar** [comment]
There is still one potential problem with the binding that I'm aware of, but don't quite know how to solve: `SchedulerImpl.schedule()` in Core uses the default timezone when calculating when to run a schedule, but this binding uses the configured timezone when making the schedules. I'm not entirely sure what the consequence is if the configured timezone is different from the OS timezone on the host. I think it will run at the wrong time, but I'm not completely sure - the whole thing is a bit convoluted, and it would require testing to establish 100% how it will play out.

It's probably not the biggest of deals, but it could make for some errors in the time between "the two different midnights", because the binding recalculates a lot of stuff at midnight to make everything apply "for the current day".

File: bundles/org.openhab.binding.astro/src/test/java/org/openhab/binding/astro/internal/model/SunTest.java [hunk 4]
```diff
@@ -49,33 +49,23 @@ public void init() {
...
     @Test
     public void testGetStateWhenNotNullPhaseName() throws Exception {
         sun.getPhase().setName(SunPhaseName.DAYLIGHT);
 
         assertEquals(new StringType("DAYLIGHT"),
-                PropertyUtils.getState(new ChannelUID("astro:sun:home:phase#name"), config, sun, ZONE));
-    }
-
-    @Test
-    public void testGetStateWhenNullPhase() throws Exception {
```
Comment (lsiepel) [inline bundles/org.openhab.binding.astro/src/test/java/org/openhab/binding/astro/internal/model/SunTest.java:72]: Can you elaborate on why this test was removed?

File: bundles/org.openhab.binding.astro/src/test/java/org/openhab/binding/astro/internal/job/JobTest.java [hunk 5]
```diff
@@ -81,20 +151,21 @@ public void adjustRangeToConfigTestSkipForceTest() {
...
         config.forceEvent = false;
-        Calendar pointInTime = DateTimeUtils.createCalendarForToday(12, 0);
+        Calendar pointInTime = DateTimeUtils.createCalendarForToday(12, 0, TIME_ZONE, Locale.ROOT);
         Range startNull = new Range(null, pointInTime);
         Range endNull = new Range(pointInTime, null);
         Range bothNull = new Range(null, null);
-        Range bothNNShouldCorrect = new Range(DateTimeUtils.createCalendarForToday(6, 0),
-                DateTimeUtils.createCalendarForToday(22, 0));
+        Range bothNNShouldCorrect = new Range(DateTimeUtils.createCalendarForToday(6, 0, TIME_ZONE, Locale.ROOT),
+                DateTimeUtils.createCalendarForToday(23, 10, TIME_ZONE, Locale.ROOT));
```
Comment (lsiepel) [inline bundles/org.openhab.binding.astro/src/test/java/org/openhab/binding/astro/internal/job/JobTest.java:159]: While reviewing i do not have a full overview of all changes yet. But did some logic change that required this time to be altered? In general i would assume that theses tests should not have values changed unless a bug was fixed.

Comment (Nadahar) [reply bundles/org.openhab.binding.astro/src/test/java/org/openhab/binding/astro/internal/job/JobTest.java:159]: I changed the values to make the test better, because the old values coincided with the "limit" used, so it would hide if the upper limit wasn't properly enforced. The limits are 08:00 and 22:00.

File: bundles/org.openhab.binding.astro/src/test/java/org/openhab/binding/astro/internal/model/SunTest.java (hunk 4 above)
Comment (Nadahar) [reply bundles/org.openhab.binding.astro/src/test/java/org/openhab/binding/astro/internal/model/SunTest.java:72]: After the `@NonNullByDefault` changes, it can no longer be `null`. So, the test is simply impossible, the code won't compile.
